core/node/provider.go

```diff
@@ -159,32 +155,65 @@ func OfflineProviders() fx.Option {
 	return fx.Provide(provider.NewNoopProvider)
 }
 
-func newProvidingStrategy(onlyPinned, onlyRoots bool) interface{} {
+func mfsProvider(mfsRoot *mfs.Root, fetcher fetcher.Factory) provider.KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		rootNode, err := mfsRoot.GetDirectory().GetNode()
+		if err != nil {
+			return nil, fmt.Errorf("error loading mfs root, cannot provide MFS: %w", err)
+		}
+
+		kcf := provider.NewDAGProvider(rootNode.Cid(), fetcher)
+		return kcf(ctx)
+	}
+
+}
+
+func mfsRootProvider(mfsRoot *mfs.Root) provider.KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		rootNode, err := mfsRoot.GetDirectory().GetNode()
+		if err != nil {
+			return nil, fmt.Errorf("error loading mfs root, cannot provide MFS: %w", err)
+		}
+		ch := make(chan cid.Cid, 1)
+		ch <- rootNode.Cid()
+		close(ch)
+		return ch, nil
+	}
+}
+
+func newProvidingStrategy(strategy string) interface{} {
 	type input struct {
 		fx.In
-		Pinner      pin.Pinner
-		Blockstore  blockstore.Blockstore
-		IPLDFetcher fetcher.Factory `name:"ipldFetcher"`
+		Pinner               pin.Pinner
+		Blockstore           blockstore.Blockstore
+		IPLDFetcher          fetcher.Factory `name:"ipldFetcher"`
+		OfflineUnixFSFetcher fetcher.Factory `name:"offlineUnixfsFetcher"`
+		MFSRoot              *mfs.Root
 	}
 	return func(in input) provider.KeyChanFunc {
-		// Pinner-related CIDs will be buffered in memory to avoid
-		// deadlocking the pinner when the providing process is slow.
-
-		if onlyRoots {
-			return provider.NewBufferedProvider(
-				provider.NewPinnedProvider(true, in.Pinner, in.IPLDFetcher),
+		switch strategy {
+		case "roots":
+			return provider.NewPrioritizedProvider(
+				provider.NewBufferedProvider(provider.NewPinnedProvider(true, in.Pinner, in.IPLDFetcher)),
+				mfsRootProvider(in.MFSRoot),
 			)
-		}
-
-		if onlyPinned {
-			return provider.NewBufferedProvider(
-				provider.NewPinnedProvider(false, in.Pinner, in.IPLDFetcher),
+		case "pinned":
+			return provider.NewPrioritizedProvider(
+				provider.NewBufferedProvider(provider.NewPinnedProvider(false, in.Pinner, in.IPLDFetcher)),
```

Comment:
Since now we can nicely traverse dags in offline mode with the offline fetcher... Shouldn't the pinned provider use an the offline fetcher for safety?

Normally pins are only written upon fetching the full merkledag, but maybe there is a way of fucking this up and having incomplete pin-dags that are not fully available locally.

Reply:
Yep, we should use offline one for safety (iiuc we already switched MFS provider related to [`Pinning.RemoteServices: Policies.MFS`](https://github.com/ipfs/kubo/blob/master/docs/config.md#pinningremoteservices-policiesmfs)). 

(If any of the local blocks is corrupted, that could trigger network fetch, which we dont want during provide-only code paths)